In [3]:
import requests

import pandas as pd

Basic stats. Let's understand the dataset better
Write about your choices in data cleaning and preprocessing
Write a short section that discusses the dataset stats, containing key points/plots from your exploratory data analysis.
Data Analysis.
Describe your data analysis and explain what you've learned about the dataset.
If relevant, talk about your machine-learning.
Genre.
Which genre of data story did you use?
Which tools did you use from each of the 3 categories of Visual Narrative (Figure 7 in Segal and Heer). Why?
Which tools did you use from each of the 3 categories of Narrative Structure (Figure 7 in Segal and Heer). Why?
Visualizations.
Explain the visualizations you've chosen.
Why are they right for the story you want to tell?



# Explainer Notebook - Final Project
02806 Social Data Analysis and Visualization - Spring 2026 - Group 42 

## Motivation
<!-- What is your dataset? -->
<!-- Why did you choose this/these particular dataset(s)? -->
<!-- What was your goal for the end user's experience? -->

### The data sets
The aim of this analysis is to showcase if socioeconomic hardship is correlated to a rise in criminal activity. The hypothesis is, that areas with a higher hardship will have a tendency to a higher crime rate, specifically in the sector of "blue-collar" crime (aka street crime). To make that analysis we will be zooming in on the city of Chicago, US, as an example. We are employing two data sets from, both collected by the city itself, to explore the possible connection between crime and socioeconomic status. 

### Why this data?
The first data set, referred to as the "socioeconomic" data set was chosen for the clear distinction of various social factors in different parts of Chicago. This data set will allow us to analyze the needed factors and compare them to a geolocation.

The other data set, referred to as the "crime" data set was chosen to showcase the crime landscape in Chicago. This data set also has access to geolocation data, which is based on the same locational categories as the ones seen in the socioeconomic data set. Since both data sets are collected by the authorities, they fit nicely together (in terms of format), and have a high level of credibility.

One caveat about the data is, that the socioeconomic data was last collected in the time period of 2008-2012 making it rather old. This analysis is therefore limited to that era.

### Main goal of analysis
We are hoping that the collected visualizations of these data sets will create an immersive journey for the reader in which they will be intrigued, not only by the correlations between hardship and crime, but by the world of visual data.

## Basic Stats
<!-- Basic stats. Let's understand the dataset better -->
<!-- Write about your choices in data cleaning and preprocessing -->
<!-- Write a short section that discusses the dataset stats, containing key points/plots from your exploratory data analysis. -->

<!-- Basic stats. Let's understand the dataset better -->

### Socioeconomic data
Find the data set at [City of Chicago - social](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2/about_data).

The first data set for this project is the Census data set collected by the city of Chicago. The aim of this data set is to document and explore the tendencies of poverty, employment, age, and education in the different communities of Chicago. In the data set we also see a custom "Hardship" score, which is an aggregation of the other parameters normalized by the area of the community.

Now, let's have a look at a sample of the data

In [ ]:
sd = pd.read_csv('./data/sociodata.csv')
print(f"{len(sd)} rows, {len(sd.columns)} columns")
# sd.describe()
sd.sample(frac=1).reset_index(drop=True).head(5)

78 rows, 9 columns


,Community Area Number,COMMUNITY AREA NAME,PERCENT OF HOUSING CROWDED,PERCENT HOUSEHOLDS BELOW POVERTY,PERCENT AGED 16+ UNEMPLOYED,PERCENT AGED 25+ WITHOUT HIGH SCHOOL DIPLOMA,PERCENT AGED UNDER 18 OR OVER 64,PER CAPITA INCOME,HARDSHIP INDEX
0,76.0,O'Hare,3.6,15.4,7.1,10.9,30.3,25828,24.0
1,73.0,Washington Height,1.1,16.9,20.8,13.7,42.6,19713,48.0
2,32.0,Loop,1.5,14.7,5.7,3.1,13.5,65526,3.0
3,23.0,Humboldt park,14.8,33.9,17.3,35.4,38.0,13781,85.0
4,48.0,Calumet Heights,2.1,11.5,20.0,11.0,44.0,28887,38.0


The data sample allows us to have an early, rough, look at the data. We can see that, except for the location, all data is in a numeric format. Most columns are presented as percentage, and thus in the range of 0 - 100. The income column is a bit different using an integer-based value and a much different range. There are 78 rows, one for each community area (and a "total" column).

#### Preprocessing
<!-- Write about your choices in data cleaning and preprocessing -->
Running through the data set there are a few things that spring to mind in terms of preprocessing. Firstly, there seem to be some spelling errors in the names of the names of the community areas, e.g. "Montclare" has been named "Montclaire" by accident. Additionally, and this is perhaps the most dangerous; the column for "PER CAPITA INCOME " has a trailing space, which can complicate some debugging down the line. Apart from that there are no errors or inconsistencies in the data. No missing values either.

#### Data set discussion
<!-- Write a short section that discusses the dataset stats, containing key points/plots from your exploratory data analysis. -->
As seen in the example we have data about the various socioeconomic factors explained previously, as well as the geographic community area. This gives an obvious opportunity to visualize the data as a choropleth plot. This visualization will allow us to analyze where, in geographical terms, there is high "hardship" in the city, as well as easily look at correlations between the different factors. This helps answer the "Where is the hardship in Chicago" part of the question we are ultimately asking.

#### What have we learned from the data?
<!-- Describe your data analysis and explain what you've learned about the dataset.Describe your data analysis and explain what you've learned about the dataset. -->
The plots helps us learn that the hardship is mostly centered in the south-western regions of the city. The coastline, especially the northern part, seems to be where the money is at (literally). The rough outline is that the more rural areas of the city are riddled with high poverty rate, low employment, and low income, while the downtown areas have a much lower hardship. Another clear fact springing to mind is just how high the poverty is, with areas reaching a per-household poverty of 56%.

### Crime data

Find this data set at [City of Chicago - crime](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-Present/ijzp-q8t2/about_data).

The second data set is the crime data, which conveniently is also collected by the city of Chicago. The aim of this data set is to explore the landscape of crime in the city, much like we have been doing throughout this course with the city of San Francisco. In difference to the socioeconomic data set, this set is quite a bit larger and features a few more columns of useful values. Not only do we get detailed infomation on the type of crime, but precise geolocation, community areas, timestamp, etc.

Let's draw a sample here as well:

In [7]:
cd = pd.read_csv('./data/crimedata.csv')
print(f"{len(cd)} rows, {len(cd.columns)} columns")
# cd.describe()
cd.sample(frac=1).reset_index(drop=True).head(5)

1878569 rows, 22 columns


,ID,Case Number,Date,Block,IUCR,Primary Type,Description,Location Description,Arrest,Domestic,...,Ward,Community Area,FBI Code,X Coordinate,Y Coordinate,Year,Updated On,Latitude,Longitude,Location
0,7901551,HT131593,01/22/2011 10:00:00 PM,088XX S JUSTINE ST,0820,THEFT,$500 AND UNDER,STREET,False,False,...,21.0,71.0,06,1167548.0,1846142.0,2011,02/04/2016 06:33:39 AM,41.733358,-87.661777,"(41.733358122, -87.661777418)"
1,7920270,HT151080,02/07/2011 01:05:00 PM,002XX E ERIE ST,1320,CRIMINAL DAMAGE,TO VEHICLE,STREET,False,False,...,42.0,8.0,14,1178076.0,1904830.0,2011,02/10/2018 03:50:01 PM,41.894171,-87.621427,"(41.894171148, -87.621427213)"
2,6283375,HP372818,06/03/2008 02:00:00 AM,023XX E 70TH PL,0890,THEFT,FROM BUILDING,RESIDENCE,False,True,...,5.0,43.0,06,1192748.0,1858661.0,2008,09/07/2021 03:41:02 PM,41.767135,-87.569051,"(41.76713466, -87.569051253)"
3,6820191,HR229164,03/22/2009 10:40:00 AM,107XX S WENTWORTH AVE,141C,WEAPONS VIOLATION,UNLAWFUL USE OTHER DANG WEAPON,SIDEWALK,True,False,...,34.0,49.0,15,1176846.0,1833909.0,2009,02/04/2016 06:33:39 AM,41.699585,-87.628082,"(41.699584978, -87.628081503)"
4,6179465,HP268420,04/08/2008 10:30:00 AM,098XX S DOBSON AVE,1330,CRIMINAL TRESPASS,TO LAND,RESIDENCE,False,True,...,8.0,50.0,26,1184954.0,1839847.0,2008,09/07/2021 03:41:02 PM,41.715694,-87.598208,"(41.715693548, -87.598208164)"


In this view we see the juice of the crime data; coordinates, precise timestamps, crime descriptions, community areas (the same as in the socioeconomic data set). We also see that there is a whooping 1,878,569 rows and 22 columns. 

#### Preprocessing
<!-- Write about your choices in data cleaning and preprocessing -->

#### Data set discussion
<!-- Write a short section that discusses the dataset stats, containing key points/plots from your exploratory data analysis. -->


#### What have we learned from the data?
<!-- Describe your data analysis and explain what you've learned about the dataset.Describe your data analysis and explain what you've learned about the dataset. -->


## Genre

<!-- Which genre of data story did you use? -->
<!-- Which tools did you use from each of the 3 categories of Visual Narrative (Figure 7 in Segal and Heer). Why? -->
<!-- Which tools did you use from each of the 3 categories of Narrative Structure (Figure 7 in Segal and Heer). Why? -->

## Visualizations
<!-- Explain the visualizations you've chosen. -->
<!-- Why are they right for the story you want to tell? -->

### Choropleth plots (Socioeconomic data set)

The map-plots were chosen to create a visual connection between the geographical data and the numeric. While possible, simply displaying the socioeconomic factors in a bar chart / histogram would not have given the same immersive effect. This type of plot instead allows the user to get a feel for where the issue exists. What exact citizens suffer under hardship? Additionally, these plots align well the geographic data in the crime data set and it presented a good opportunity for a qualitative assessment of the correlation between the two data sets.

In [ ]:
def _get_shapes_file() -> dict:
    """Return geojson for chicago community areas"""
    # import shapefile for community areas
    shapefile = requests.get("https://data.cityofchicago.org/resource/igwz-8jzy.json").json()

    # And convert them to proper geojson format
    geojson = {
        "type": "FeatureCollection",
        "features": [
            {
                "type": "Feature",
                "geometry": row["the_geom"],
                "properties": {k: v for k, v in row.items() if k != "the_geom"},
            }
            for row in shapefile
            if row.get("the_geom") and row["the_geom"].get("coordinates")
        ],
    }
    return geojson

def _align_naming_schemes(df: pd.DataFrame) -> pd.DataFrame:
    """Align the naming schemes of the comminity areas (city of chicago misspelled some names in the census data... smh)"""
    df = df.copy()
    # Spell correction
    name_fixes = {
        "MONTCLAIRE": "MONTCLARE",
        "WASHINGTON HEIGHT": "WASHINGTON HEIGHTS",
        "O'HARE": "OHARE",
    }
    df["COMMUNITY AREA NAME"] = df["COMMUNITY AREA NAME"].str.upper().replace(name_fixes)
    # Exclude the "total" row
    df = df[df["COMMUNITY AREA NAME"] != "CHICAGO"]
    return df

def plot_income_choropleth(col: str, filename: str, plot_data: pd.DataFrame, geojson: dict) -> None:
    import plotly.express as px
    import plotly.graph_objects as go

    fig = go.Figure(
        data=px.choropleth_map(
            plot_data,
            geojson=geojson,
            locations="COMMUNITY AREA NAME",
            color=col,
            featureidkey="properties.community",
            color_continuous_scale="Magma",
            map_style="carto-positron",
            zoom=9,
            center={"lat": 41.8781, "lon": -87.6298},
            opacity=0.6,
        )
    )
    fig.update_layout(margin={"r": 0, "t": 0, "l": 0, "b": 0})
    fig.write_html(f"./../docs/figures/{filename}", include_plotlyjs="cdn")

geojson = _get_shapes_file()
plot_data = _align_naming_schemes(sd)

# Generate choropleth plots for the 4 main columns
plot_income_choropleth("HARDSHIP INDEX", "hardship_choropleth.html", plot_data, geojson)
plot_income_choropleth("PER CAPITA INCOME ", "income_choropleth.html", plot_data, geojson) # Yes, trailing space is on purpose... smh
plot_income_choropleth("PERCENT HOUSEHOLDS BELOW POVERTY", "poverty_choropleth.html", plot_data, geojson)
plot_income_choropleth("PERCENT AGED 16+ UNEMPLOYED", "unemployment_choropleth.html", plot_data, geojson)

## Discussion
<!-- Think critically about your creation -->
<!-- What went well? -->
<!-- What is still missing? What could be improved? Why? -->